### PR TITLE
add shared object.

### PR DESCRIPTION
**Pull request by vwxyzh**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Concurrent;
+
+    public class SharedObjectManager<TState, TOperand>
+    {
+        private readonly ConcurrentDictionary<TState, SharedObjectNode<TState, TOperand>> _states;
+        private readonly IEqualityComparer<TState> _stateComparer;
+        private readonly Func<TState, TOperand, TState> _modifier;
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >_modifier [](start = 56, length = 9)

_transition?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Common
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Concurrent;
+
+    public class SharedObjectManager<TState, TOperand>
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >TOperand [](start = 45, length = 8)

TCondition?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            _states = new ConcurrentDictionary<TState, SharedObjectNode<TState, TOperand>>(_stateComparer);
+            RootState = new SharedObjectNode<TState, TOperand>(this, rootState);
+            _states.TryAdd(rootState, RootState);
+        }
+
+        public SharedObjectNode<TState, TOperand> RootState { get; }
+
+        internal IEqualityComparer<TOperand> OperandComparer { get; }
+
+        internal SharedObjectNode<TState, TOperand> Modify(TState value, TOperand operand) =>
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >Modify(TState value, TOperand operand) [](start = 52, length = 38)

ChangeState(TState state, TCondition condition)?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Concurrent;
+
+    public class SharedObjectManager<TState, TOperand>
+    {
+        private readonly ConcurrentDictionary<TState, SharedObjectNode<TState, TOperand>> _states;
+        private readonly IEqualityComparer<TState> _stateComparer;
+        private readonly Func<TState, TOperand, TState> _modifier;
+        private readonly Func<TState, SharedObjectNode<TState, TOperand>> _creator;
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >_creator [](start = 74, length = 8)

can be fixed?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs (hunk 1 above)
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >_modifier [](start = 56, length = 9)

_transitionRule?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            _stateComparer = stateComparer ?? EqualityComparer<TState>.Default;
+            _modifier = modifier;
+            _creator = CreateNewNode;
+            OperandComparer = operandComparer ?? EqualityComparer<TOperand>.Default;
+            _states = new ConcurrentDictionary<TState, SharedObjectNode<TState, TOperand>>(_stateComparer);
+            RootState = new SharedObjectNode<TState, TOperand>(this, rootState);
+            _states.TryAdd(rootState, RootState);
+        }
+
+        public SharedObjectNode<TState, TOperand> RootState { get; }
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >RootState [](start = 50, length = 9)

naming: Current?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        private readonly Func<TState, SharedObjectNode<TState, TOperand>> _creator;
+
+        public SharedObjectManager(
+            TState rootState,
+            Func<TState, TOperand, TState> modifier,
+            IEqualityComparer<TState> stateComparer = null,
+            IEqualityComparer<TOperand> operandComparer = null)
+        {
+            _stateComparer = stateComparer ?? EqualityComparer<TState>.Default;
+            _modifier = modifier;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >modifier [](start = 24, length = 8)

null check

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectNode.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        {
+            _manager = obj;
+            _path = new ConcurrentDictionary<TOperand, SharedObjectNode<TState, TOperand>>(_manager.OperandComparer);
+            Value = value;
+            _creator = o => _manager.Modify(Value, o);
+        }
+
+        public TState Value { get; }
+
+        public SharedObjectNode<TState, TOperand> Modify(TOperand operand) =>
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectNode.cs]: >operand [](start = 66, length = 7)

check null?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            _creator = CreateNewNode;
+            OperandComparer = operandComparer ?? EqualityComparer<TOperand>.Default;
+            _states = new ConcurrentDictionary<TState, SharedObjectNode<TState, TOperand>>(_stateComparer);
+            RootState = new SharedObjectNode<TState, TOperand>(this, rootState);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >new SharedObjectNode [](start = 24, length = 20)

call _creator to keep consistency?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs (hunk 2 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: rename to TEvent

---
In reply to: [120027375](https://github.com/dotnet/docfx/pull/1723#discussion_r120027375) [](ancestors = 120027375)

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs (hunk 1 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: rename to _transit

---
In reply to: [120028066](https://github.com/dotnet/docfx/pull/1723#discussion_r120028066) [](ancestors = 120028066)

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs (hunk 4 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: avoid new func per call

---
In reply to: [120027775](https://github.com/dotnet/docfx/pull/1723#discussion_r120027775) [](ancestors = 120027775)

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs (hunk 5 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: this is root state.

---
In reply to: [120035824](https://github.com/dotnet/docfx/pull/1723#discussion_r120035824) [](ancestors = 120035824)

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs (hunk 3 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: update to Transit(TState value, TEvent @event)

---
In reply to: [120027642](https://github.com/dotnet/docfx/pull/1723#discussion_r120027642) [](ancestors = 120027642)

File: src/Microsoft.DocAsCode.Common/StateMachines/StateNode.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+    public class StateNode<TState, TEvent>
+    {
+        private readonly StateMachine<TState, TEvent> _manager;
+        private readonly ConcurrentDictionary<TEvent, StateNode<TState, TEvent>> _path;
+        private readonly Func<TEvent, StateNode<TState, TEvent>> _creator;
+
+        internal StateNode(StateMachine<TState, TEvent> obj, TState value)
+        {
+            _manager = obj;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/StateMachines/StateNode.cs]: >_manager [](start = 12, length = 8)

minor: rename to stateMachine?

File: src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+namespace Microsoft.DocAsCode.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+
+    public class SharedObjectManager<TState, TEvent>
+    {
+        private readonly ConcurrentDictionary<TState, SharedObjectNode<TState, TEvent>> _states;
+        private readonly IEqualityComparer<TState> _stateComparer;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/SharedObjects/SharedObjectManager.cs]: >_stateComparer [](start = 51, length = 14)

no need to private field? which is only used in ctor?

**Comment by hellosnow** [comment]
:shipit:

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**